py-polars/src/lazyframe/visitor/nodes.rs

```diff
@@ -257,29 +257,41 @@ pub struct Sink {
 
 pub(crate) fn into_py(py: Python<'_>, plan: &IR) -> PyResult<PyObject> {
     let result = match plan {
-        IR::PythonScan { options, predicate } => PythonScan {
-            options: (
-                options
-                    .scan_fn
-                    .as_ref()
-                    .map_or_else(|| py.None(), |s| s.0.clone()),
-                options
-                    .with_columns
-                    .as_ref()
-                    .map_or_else(|| py.None(), |cols| cols.to_object(py)),
-                options.pyarrow,
-                options
-                    .predicate
-                    .as_ref()
-                    .map_or_else(|| py.None(), |s| s.to_object(py)),
-                options
-                    .n_rows
-                    .map_or_else(|| py.None(), |s| s.to_object(py)),
-            )
-                .to_object(py),
-            predicate: predicate.as_ref().map(|e| e.into()),
-        }
-        .into_py(py),
+        IR::PythonScan { options } => {
+            let python_src = match options.python_source {
+                PythonScanSource::Pyarrow => "pyarrow",
+                PythonScanSource::Cuda => "cuda",
+                PythonScanSource::IOPlugin => "io_plugin",
+            };
+
+            PythonScan {
+                options: (
+                    options
+                        .scan_fn
+                        .as_ref()
+                        .map_or_else(|| py.None(), |s| s.0.clone()),
+                    options
+                        .with_columns
+                        .as_ref()
+                        .map_or_else(|| py.None(), |cols| cols.to_object(py)),
+                    python_src,
+                    match &options.predicate {
+                        PythonPredicate::None => py.None(),
+                        PythonPredicate::PyArrow(s) => s.to_object(py),
+                        PythonPredicate::Polars(_) => {
+                            return Err(PyNotImplementedError::new_err(
+                                "polars native predicates not yet supported",
+                            ))
+                        },
```

Comment:
This could be:

```
PythonPredicate::PyArrow(s) => ("pyarrow", s.to_object(py)),
PythonPredicate::Polars(e) => ("polars", e.into()),
```

(I've been using string-tagged tuples as a poor approximation of tagged enums, that way, I don't have to do isinstance checks to determine which of the two types of expression I have).

On the translation side, we know how to handle `ExprIR` objects.